drivers/nec/display/display.cr

```diff
@@ -0,0 +1,313 @@
+    DisplayPort = 15
+    Hdmi        = 17
+    Hdmi2       = 18
+    Hdmi3       = 130
+    Usb         = 135
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Input)
+
+  # Discovery Information
+  tcp_port 7142
+  descriptive_name "NEC Display"
+  generic_name :Display
+
+  DELIMITER = 0x0D_u8
+
+  def on_load
+    # Communication settings
+    queue.delay = 120.milliseconds
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+  end
+
+  def connected
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    # Do nothing if already in desired state
+    return if self[:power]?.try &.as_bool == state
+
+    if state
+      # 1 = Power On
+      data = [Command::Power.value, 1]
+      logger.debug { "-- NEC LCD, requested to power on" }
+      do_send(MsgType::Command, data, name: "power", delay: 5.seconds)
+    else
+      logger.debug { "-- NEC LCD, requested to power off" }
+      # 4 = Power Off
+      data = [Command::Power.value, 4]
+      do_send(MsgType::Command, data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+
+      self[:power] = false
+    end
+  end
+
+  def power?(**options) : Bool
+    do_send(MsgType::Command, Command::PowerQuery, **options, name: "power?").get
+    self[:power].as_bool
+  end
+
+  def switch_to(input : Input)
+    data = [Command::VideoInput.value, input.value]
+
+    logger.debug { "-- NEC LCD, requested to switch to: #{input}" }
+    do_send(MsgType::SetParameter, data, name: "input", delay: 6.seconds)
+    video_input
+  end
+
+  enum Audio
+    Audio1      = 1
+    Audio2      = 2
+    Audio3      = 3
+    Hdmi        = 4
+    Tv          = 6
+    DisplayPort = 7
+  end
+
+  def switch_audio(input : Audio)
+    data = [Command::AudioInput.value, input.value]
+
+    logger.debug { "-- NEC LCD, requested to switch audio to: #{input}" }
+    do_send(MsgType::SetParameter, data, name: "audio")
+  end
+
+  def auto_adjust
+    data = [Command::AutoSetup.value, 1]
+    do_send(MsgType::SetParameter, data, name: "auto_adjust")
+  end
+
+  def brightness(val : Int32)
+    data = [Command::BrightnessStatus.value, val.clamp(0, 100)]
+
+    do_send(MsgType::SetParameter, data, name: "brightness")
+    do_send(MsgType::Command, Command::Save, name: "brightness_save") # Save the settings
+  end
+
+  def contrast(val : Int32)
+    data = [Command::ContrastStatus.value, val.clamp(0, 100)]
+
+    do_send(MsgType::SetParameter, data, name: "contrast")
+    do_send(MsgType::Command, Command::Save, name: "contrast_save") # Save the settings
+  end
+
+  def volume(val : Int32)
+    data = [Command::VolumeStatus.value, val.clamp(0, 100)]
+
+    do_send(MsgType::SetParameter, data, name: "volume")
+    do_send(MsgType::Command, Command::Save, name: "volume_save") # Save the settings
+    self[:audio_mute] = false # audio is unmuted when the volume is set
```

Comment:
This may also be a good bit of state to update after the device has responded rather than here.